cleanrl/sac_atari.py

```diff
@@ -0,0 +1,374 @@
+            torch.nn.init.constant_(param.bias, bias_const)
+    
+    return net
+
+
+
+# ALGO LOGIC: initialize agent here:
+# NOTE: Sharing a CNN encoder between Actor and Critics is not recommended for SAC
+# See https://arxiv.org/abs/1910.01741 for more info
+# TL;DR The actor's gradients mess up the representation when using a joint encoder
+class SoftQNetwork(nn.Module):
+    """Soft Q network for pixel inputs."""
+
+    def __init__(self, envs):
+        super().__init__()
+        obs_shape = envs.single_observation_space.shape
+        self.conv = nn.Sequential(
+            nn.Conv2d(obs_shape[0], 32, kernel_size=8, stride=4),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, kernel_size=4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(64, 64, kernel_size=3, stride=1),
+            nn.Flatten(),
+            nn.ReLU(),
+        )
+
+        with torch.inference_mode():
+            output_dim = self.conv(torch.zeros(1, *obs_shape)).shape[1]
+
+        self.Q_func = nn.Sequential(
+            nn.Linear(output_dim, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n),
+        )
+
+    def forward(self, x):
+        x = self.conv(x / 255.0)
+        x = self.Q_func(x)
+        return x
+
+
+class Actor(nn.Module):
+    """Discrete SAC Actor for pixel inputs"""
+
+    def __init__(self, envs):
+        super().__init__()
+        obs_shape = envs.single_observation_space.shape
+        self.conv = nn.Sequential(
+            nn.Conv2d(obs_shape[0], 32, kernel_size=8, stride=4),
+            nn.ReLU(),
+            nn.Conv2d(32, 64, kernel_size=4, stride=2),
+            nn.ReLU(),
+            nn.Conv2d(64, 64, kernel_size=3, stride=1),
+            nn.Flatten(),
+            nn.ReLU(),
+        )
+
+        with torch.inference_mode():
+            output_dim = self.conv(torch.zeros(1, *obs_shape)).shape[1]
+
+        self.fc_logits = nn.Sequential(
+            nn.Linear(output_dim, 512),
+            nn.ReLU(),
+            nn.Linear(512, envs.single_action_space.n),
+        )
+
+        # He initialization
+        for param in self.parameters():
+            if type(param) == nn.Linear or type(param) == nn.Conv2d:
+                nn.init.kaiming_normal_(param.weight)
+
+    def forward(self, x):
+        x = self.conv(x)
+        logits = self.fc_logits(x)
+
+        return logits
+
+    def get_action(self, x):
+        logits = self(x / 255.0)
+        policy_dist = Categorical(logits=logits)
+        action = policy_dist.sample()
+        # Action probabilities for calculating the adapted soft-Q loss
+        action_probs = policy_dist.probs
+        # From https://github.com/p-christ/Deep-Reinforcement-Learning-Algorithms-with-PyTorch/blob/135d3e2e06bbde2868047d738e3fc2d73fd8cc93/agents/actor_critic_agents/SAC_Discrete.py#L58
+        # Add small numerical constant where action probability is 0 to avoid log(0)
+        z = action_probs == 0.0
```

Comment:
From my experience, you can achieve the same with `F.log_softmax(logits, dim=-1)`, which is more explicit and convenient

Reply:
Thanks for the hint! Changed it.